src/source-control/GitSourceControl.ts

```diff
@@ -1,47 +1,449 @@
-import type { Commit, CommitHash } from "../shared/types";
+import nodegit from "nodegit";
+import produce, { enableMapSet } from "immer";
+import fs from "fs";
+import { lorem } from "faker";
+enableMapSet();
+
+import type {
+  BranchName,
+  Commit,
+  CommitHash,
+  Repository,
+  CommitSignature,
+} from "../shared/types";
 import type {
   SourceControl,
   SourceControlRepositoryUpdateListener,
 } from "./SourceControl";
 
+const localRefPrefix: string = "refs/heads/";
+
+function refIsLocal(refString: string): boolean {
+  return refString.startsWith(localRefPrefix);
+}
+
+function localRefToBranchName(refString: string): string {
+  if (!refIsLocal(refString)) return refString;
+  return refString.substr(localRefPrefix.length);
+}
+
+function isSttackBranch(branch: BranchName): boolean {
+  return branch.startsWith("sttack-");
+}
+
+function createSttackBranch(): BranchName {
+  const branchName: BranchName = `sttack-${lorem.slug(3)}`;
+  return branchName;
+}
+
 export class GitSourceControl implements SourceControl {
   private repoPath: string;
 
   repositoryUpdateListener: SourceControlRepositoryUpdateListener;
+  private commitHashMap = new Map<string, Commit>();
 
   constructor(
     repoPath: string,
     repositoryUpdateListener: SourceControlRepositoryUpdateListener,
   ) {
     this.repoPath = repoPath;
     this.repositoryUpdateListener = repositoryUpdateListener;
+    this.populateGitSourceControl();
```

Comment:
Oh this is pretty dangerous, as you may end up trying to use the data before it’s ready. Could we avoid this?

Reply:
How do you suggest I go about that?

Reply:
The cleanest option I can think of is to add a private function like `private async getGuaranteedPopulatedData` or similar, that either returns cached data if it exists or loads new data. You can then use that everywhere you need preloaded data.

Feel free to think of other options, but this is the best one I can think of right now

Reply:
Implemented an alternate, let me know if you think it's okay